clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/ActiveActiveStoreIngestionTask.java

```diff
@@ -298,35 +298,55 @@ protected void removeFromStorageEngine(int partition, byte[] keyBytes, Delete de
     }
   }
 
-  /** @return what kind of storage operation to execute, if any. */
-  private StorageOperationType getStorageOperationType(int partition, ByteBuffer valuePayload, ByteBuffer rmdPayload) {
-    PartitionConsumptionState pcs = partitionConsumptionStateMap.get(partition);
+  /**
+   * This method determines which type of storage operation it is based on input value and RMD payload.
+   */
+  StorageOperationType getStorageOperationType(int partition, ByteBuffer valuePayload, ByteBuffer rmdPayload) {
+    PartitionConsumptionState pcs = getPartitionConsumptionStateMap().get(partition);
     if (pcs == null) {
       logStorageOperationWhileUnsubscribed(partition);
-      return StorageOperationType.NONE;
-    }
-    if (isDaVinciClient) {
-      return StorageOperationType.VALUE;
+      return StorageOperationType.SKIP;
     }
     if (rmdPayload == null) {
       throw new IllegalArgumentException("Replication metadata payload not found.");
     }
-    if (pcs.isEndOfPushReceived() || rmdPayload.remaining() > 0) {
-      // value payload == null means it is a DELETE request, while value payload size > 0 means it is a PUT request.
-      if (valuePayload == null || valuePayload.remaining() > 0) {
-        return StorageOperationType.VALUE_AND_RMD;
+    if (valuePayload != null && valuePayload.remaining() == 0 && rmdPayload.remaining() == 0) {
+      throw new IllegalArgumentException("Both value and RMD payload cannot be empty.");
```

Comment:
Do you mean at least one of `value` or `rmd` payload should carry a valid value? 
The log message is a little confusing to me since it indicates both `value` and rmd` should carry valid values all the time.
If we break down the logic into two branches as mentioned in the previous comment, it will be more clear.